sdk/pinocchio/src/sysvars/slot_hashes/raw.rs

```diff
@@ -0,0 +1,111 @@
+//! Raw / caller-supplied buffer helpers for the SlotHashes sysvar.
+//!
+//! This sub-module exposes lightweight functions that let a program copy
+//! SlotHashes data directly into an arbitrary buffer **without** constructing
+//! a `SlotHashes<T>` view. Use these when you only need a byte snapshot or
+//! when including the sysvar account is infeasible.
+#![allow(clippy::inline_always)]
+
+use super::*;
+
+/// Validates that a buffer is properly sized for SlotHashes data.
+///
+/// Checks that the buffer length is 8 + (N × 40) for some N ≤ 512.
+/// Unlike the `SlotHashes` constructor, this function does not require
+/// the buffer to be exactly 20,488 bytes.
+#[inline(always)]
+pub(crate) fn validate_buffer_size(buffer_len: usize) -> Result<(), ProgramError> {
+    // Must have space for 8-byte header
+    if buffer_len < NUM_ENTRIES_SIZE {
+        return Err(ProgramError::AccountDataTooSmall);
+    }
+
+    // Calculate how many entries can fit
+    let data_len = buffer_len - NUM_ENTRIES_SIZE;
+    if data_len % ENTRY_SIZE != 0 {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let max_entries = data_len / ENTRY_SIZE;
+    if max_entries > MAX_ENTRIES {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    Ok(())
+}
+
+/// Validates offset parameters for fetching SlotHashes data.
+///
+/// * `offset` – Byte offset within the SlotHashes sysvar data.
+/// * `buffer_len` – Length of the destination buffer.
+#[inline(always)]
+pub fn validate_fetch_offset(offset: usize, buffer_len: usize) -> Result<(), ProgramError> {
+    if offset >= MAX_SIZE {
+        return Err(ProgramError::InvalidArgument);
+    }
+    if offset != 0 && (offset < NUM_ENTRIES_SIZE || (offset - NUM_ENTRIES_SIZE) % ENTRY_SIZE != 0) {
+        return Err(ProgramError::InvalidArgument);
+    }
+    if offset.saturating_add(buffer_len) > MAX_SIZE {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    Ok(())
+}
+
+/// Copies SlotHashes sysvar bytes into `buffer`, performing validation.
+///
+/// Returns the number of entries present in the sysvar.
+#[inline(always)]
+pub fn fetch_into(buffer: &mut [u8], offset: usize) -> Result<usize, ProgramError> {
+    if buffer.len() != MAX_SIZE {
+        validate_buffer_size(buffer.len())?;
+    }
+
+    validate_fetch_offset(offset, buffer.len())?;
+
+    fetch_into_unchecked(buffer, offset)?;
+
+    let num_entries = read_entry_count_from_bytes(buffer).unwrap_or(0);
+
+    // Reject oversized entry counts to prevent surprises.
+    if num_entries > MAX_ENTRIES {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let required_len = NUM_ENTRIES_SIZE + num_entries * ENTRY_SIZE;
+    if buffer.len() < required_len {
+        return Err(ProgramError::InvalidArgument);
+    }
```

Comment:
I am not sure the need for this checks. I would assume that the syscall did the right thing, so `num_entries` would never be greater than `MAX_ENTRIES` and if the length was invalid, the syscall would fail.

Reply:
removed `num_entries > MAX_ENTRIES` check in 551b0f9

Reply:
Do we need the `if buffer.len() < required_le` check? This is after the syscall was called, so if the buffer length was invalid, the syscall would have failed.

Reply:
I'm glad you pointed this out; as I double-checked the context, I noticed that the code was returning incorrect entry counts when `offset > 0`.

The fix is a rewrite of `fetch_into` and `validate_buffer_size`, plus updated tests. Done in 89d21ab